Tests/AlicerceTests/AutoLayout/ConstraintGroupToggleTestCase.swift

```diff
@@ -0,0 +1,193 @@
+    }
+
+    func testActivate_WithSecondAndThirdConstraint_ShouldDeactivateFirstAndSecondConstraintAndActivateThirdConstraint() {
+
+        constraintGroupToggle.activate(.second)
+
+        XCTAssertFalse(constraint1.isActive)
+        XCTAssert(constraint2.isActive)
+        XCTAssertFalse(constraint3.isActive)
+
+        constraintGroupToggle.activate(.third)
+
+        XCTAssertFalse(constraint1.isActive)
+        XCTAssertFalse(constraint2.isActive)
+        XCTAssert(constraint3.isActive)
+    }
+
+    func testActivate_WithSecondAndFirstConstraint_ShouldDeactivateSecondAndActivateFirstConstraint() {
+
+        constraintGroupToggle.activate(.second)
+
+        XCTAssertFalse(constraint1.isActive)
+        XCTAssert(constraint2.isActive)
+        XCTAssertFalse(constraint3.isActive)
+
+        constraintGroupToggle.activate(.first)
+
+        XCTAssertTrue(constraint1.isActive)
+        XCTAssertFalse(constraint2.isActive)
+        XCTAssertFalse(constraint3.isActive)
+    }
+
+    func testActivate_WithUnhandledKey_ShouldDeactivateAllConstraints() {
+
+        let nonFiniteConstraintGroupToggle = ConstraintGroupToggle(
+            initial: "constraint1",
+            constraintGroups: ["constraint1": constraintGroup2, "constraint2": constraintGroup3]
+        )
+
+        nonFiniteConstraintGroupToggle.activate("constraint3")
+
+        XCTAssertFalse(constraint2.isActive)
+        XCTAssertFalse(constraint3.isActive)
+    }
+
+    func testActivate_WithSecondConstraint_ShouldDeactivatePreviouConstraintGroupBeforeActivatingSecond() {
+
+        let activateClosureOne = self.expectation(description: "ConstraintGroup one activated")
+        let deactivateClosureOne = self.expectation(description: "ConstraintGroup one deactivated")
+        let activateClosureTwo = self.expectation(description: "ConstraintGroup two activated")
+
+        let one = MockContraintGroup(
+            activateClosure: {
+                activateClosureOne.fulfill()
+            },
+            deactivateClosure: {
+                deactivateClosureOne.fulfill()
+            }
+        )
+
+        let two = MockContraintGroup(
+            activateClosure: {
+                activateClosureTwo.fulfill()
+            },
+            deactivateClosure: {
+                XCTFail()
+            }
+        )
+
+        let nonFiniteConstraintGroupToggle = ConstraintGroupToggle(
+            initial: "constraint1",
+            constraintGroups: ["constraint1": one, "constraint2": two]
+        )
+
+        nonFiniteConstraintGroupToggle.activate("constraint2")
+
+        wait(
+            for: [activateClosureOne, deactivateClosureOne, activateClosureTwo],
+            timeout: 1,
+            enforceOrder: true
+        )
+    }
+}
+
+private enum TestConstraintGroupKey: Hashable {
+    case first
+    case second
+    case third
+}
+
+private final class MockContraintGroup: ConstraintGroup {
+
+    private let activateClosure: () -> Void
```

Comment:
no need to be so concerned with encapsulation in mocks, and we don't need to pass them on `init`, as it makes them more cumbersome to use.

IMO this could be a single optional `var` closure like `var didSetIsActive: ((Bool) -> Void)?`, which we then invoke on the `isActive` property `didSet`. This allows users to use mocks easily and only intercept calls when necessary.

We also try to avoid that mocks have "business logic", like `isConstraintActive` and private functions.

Does it make sense?

Reply:
As discussed, `ConstraintGroup`'s internal mechanics make it harder to mock and test, and as such let's leave that problem for the future 😉